services/src/main/java/org/keycloak/authentication/authenticators/browser/ConditionalSpnegoAuthenticator.java

```diff
@@ -0,0 +1,87 @@
+/*
+ * Copyright 2023 Red Hat, Inc. and/or its affiliates
+ * and other contributors as indicated by the @author tags.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+package org.keycloak.authentication.authenticators.browser;
+
+import org.jboss.logging.Logger;
+import org.keycloak.authentication.AuthenticationFlowContext;
+import org.keycloak.sessions.AuthenticationSessionModel;
+
+import java.util.Map;
+import java.util.regex.Pattern;
+
+/**
+ * A {@link SpnegoAuthenticator} that can conditionally skip SPNEGO authentication.
+ * <p>
+ * This Keycloak authenticator allows SPNEGO authentication to be skipped by explicit client request and optionally by
+ * request IP. Clients can request SPNEGO authentication be skipped via the URL prompt=login present in the client
+ * request. A Whitelist Regex Pattern can optionally be configured, and if configured, it is compared against the
+ * <i>X-Forwarded-For</i> header on auth requests to check for a match. If no match then no SPNEGO is skipped.
+ * </p>
+ * <p>
+ * It is often desirable to treat Intranet (internal) users with the convenience of Kerberos Single Sign On, while
+ * preventing Internet (external) users from possibly seeing a confusing NTLM prompt dialog.  It can also be useful to
+ * allow users to explicitly opt out of SPNEGO in order to provide a switch user functionality.
+ * </p>
+ * @author <a href="mailto:Ryan.Slominski@gmail.com">Ryan Slominski</a>
+ */
+public class ConditionalSpnegoAuthenticator extends SpnegoAuthenticator {
+
+    private static Logger logger = Logger.getLogger(ConditionalSpnegoAuthenticator.class);
+
+    public static final String WHITELIST_PATTERN = "XForwardedForWhitelistPattern";
+
+    @Override
+    public void authenticate(AuthenticationFlowContext context) {
+        AuthenticationSessionModel session = context.getAuthenticationSession();
+        Map<String, String> clientNotes = session.getClientNotes();
+        if ("login".equals(clientNotes.get("prompt"))) {
```

Comment:
Few points to this:
 
1) Is it possible to replace this with something like:
```
LoginProtocol protocol = context.getSession().getProvider(LoginProtocol.class, authSession.getProtocol());
if (protocol.requireReauthentication(context.getUserSession(), context.getAuthenticationSession());
```

Will make it possible to use this with context of SAML as well. It is possible that this will need to add method `getUserSession()` to `AuthenticationFlowContext` (UserSession might have been attached before by CookieAuthenticator, which is usually before SPNEGO).

Also will likely require some checks for null userSession in the protocol implementations.

2) Is it possible to add option to the condition like `Ignore when protocol requires re-authentication` with the description similar to `When this option is enabled, then condition is evaluated to false when protocol requires re-authentication. For example when parameter prompt=login for OIDC protocol is used or forceauthn for SAML protocol` .

This will allow that condition can be evaluated for:
- only protocol authentication
- only x-forwarded header
- both

Reply:
It sounds like we may need two new separate `ConditionalAuthenticator`s:

1. XForwardedForPatternMatcherCondition
2.  PromptLoginCondition

Reply:
Ah, I see, yes supporting both SAML and OIDC would be best.   So instead of PromptLoginCondition perhaps we call it ReauthenticationCondition.

Reply:
yes, 2 separate conditions can be better option. +1 for `ReauthenticationCondition` (will be good to add option to it which allows to evaluate it to true in case that protocol re-authentication is used but also for the opposite case when protocol re-authentication is *not* used. As I think that for various cases, both conditions can be useful).

Reply:
> but also for the opposite case when protocol re-authentication is not used

This sounds like something the `ConditionalAuthenticator` framework should support generically - i.e. a config option to apply `not` (!) on the boolean result that a condition provides.

Reply:
> This sounds like something the `ConditionalAuthenticator` framework should support generically - i.e. a config option to apply `not` (!) on the boolean result that a condition provides.

+1, it does not right now, but it is generally useful for every condition though.